pkg/virt-controller/watch/vm_initializer.go

```diff
@@ -0,0 +1,217 @@
+package watch
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/api/core/v1"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	kubev1 "kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+type VMInitializer struct {
+	clientset  kubecli.KubevirtClient
+	restClient *rest.RESTClient
+	queue      workqueue.RateLimitingInterface
+	store      cache.Store
+}
+
+const initializerMarking = "virtualmachines.kubevirt.io"
+const annotationPrefix = "virtualmachinepreset.admission.kubernetes.io/virtualmachinepreset"
+
+// FIXME: Both the restClient and clientSet are probably not needed.
+func NewVMInitializer(restClient *rest.RESTClient, queue workqueue.RateLimitingInterface, vmCache cache.Store, clientset kubecli.KubevirtClient) *VMInitializer {
+	vmi := VMInitializer{
+		restClient: restClient,
+		clientset:  clientset,
+		store:      vmCache,
+		queue:      queue,
+	}
+	return &vmi
+}
+
+func (c *VMInitializer) initializeVM(vm *kubev1.VirtualMachine) error {
+	// All VM's must be marked as initialized or they are held in limbo forever
+	// Collect all errors and defer returning until after the update
+	logger := log.Log
+	errors := []error{}
+
+	logger.Object(vm).Info("Initializing VM")
+
+	allPresets, err := listPresets(c.restClient, vm.GetNamespace(), labels.Everything())
+	//list, err := c.clientset.VirtualMachinePresets(a.GetNamespace()).List(labels.Everything())
+
+	matchingPresets, err := filterPresets(allPresets, vm)
+
+	if err != nil {
+		logger.Object(vm).Reason(err).Errorf("Error while matching presets to VirtualMachine")
+		errors = append(errors, err)
+	}
+
+	if len(matchingPresets) != 0 {
+		err = checkPresetMergeConflicts(vm, matchingPresets)
+
+		if err != nil {
+			logger.Reason(err).Errorf("Conflicting VM Presets")
+			errors = append(errors, fmt.Errorf("conflicting vm presets"))
+		}
+
+		err = applyPresets(vm, matchingPresets)
+		if err != nil {
+			// A more specific error should have been logged during the applyPresets call.
+			// We don't know *which* preset in the list was problematic at this level.
+			logger.Object(vm).Errorf("Unable to apply presets to virtualmachine: %v", err)
+			errors = append(errors, err)
+		}
+	}
+
+	logger.Object(vm).Info("Marking VM as initialized and updating")
+	removeInitializer(vm)
+	c.clientset.VM(vm.Namespace).Update(vm)
+
+	if len(errors) > 0 {
+		utilerrors.NewAggregate(errors)
+	}
+	return nil
+}
+
+// FIXME: this should be integrated into the KubeVirt clientset
+func listPresets(restClient *rest.RESTClient, namespace string, selector labels.Selector) ([]kubev1.VirtualMachinePreset, error) {
```

Comment:
Please create a ListWatcher and and look the presets up from a cache. That is also what the pod presets are doing. The lister [here](https://github.com/kubernetes/kubernetes/blob/ff58401257526da1aa4c22b674fe1956e20c9c7b/pkg/client/listers/settings/internalversion/podpreset.go#L77) is used there and that only talks to the cache.

Reply:
I've been meaning to circle back and get rid of that function anyways, but a ListWatcher is a fantastic approach.